src/utils/miscellaneous.ts

```diff
@@ -217,15 +218,14 @@ export interface TimeoutParams<T> {
   timeoutErrorMsg?: string
 }
 
-export function timeout<T>(params: TimeoutParams<T>): Promise<T> {
+export function timeout(params: TimeoutParams<undefined>): Promise<never> // never means function throws
+export function timeout<T>(params: TimeoutParams<T extends undefined ? never : T>): Promise<T>
+export async function timeout<T>(params: TimeoutParams<T>): Promise<T | never> {
   const { time = DEFAULT_TIMEOUT, result, timeoutErrorMsg: timeoutMsg = 'Timeout' } = params
-  return new Promise((resolve, rejects) =>
-    setTimeout(() => {
-      if (result) {
-        resolve(result)
-      } else {
-        rejects(new Error(timeoutMsg))
-      }
-    }, time),
-  )
+
+  await delay(time)
+  // provided defined result -- return it
+  if (result !== undefined) return result
+  // no defined result -- throw message
+  throw timeoutMsg
```

Comment:
why discard the `Error` constructor? guess it is functionally the same?

Reply:
No you are right, should be an `Error`